### ISSUE TITLE
Verifier Finetuning with Generated train dataset

### ISSUE THREAD

**Issue by chenhan97** [FreedomIntelligence/OVM#5]
Hi, thank you for your work and the released code!

I've encountered an issue when training the verifier and evaluating "generator+verifier" (the eval_step_beam.sh). I hope you may help find where is the mistake.

I first trained the generator following the instructions, namely "train_generator.sh", to obtain a generator Mistral-A.

Then I trained a verifier model using the following script and the dataset provided (https://huggingface.co/datasets/FreedomIntelligence/OVM-dataset/blob/main/gsm8k/model_generation/train/responses_n100_mistral7b-ep2.jsonl), namely Mistral-B.

here is the script that i'm using for training the verifier, (basically the same as the provided script), where n_solution=100:
```
accelerate launch \
  --config_file ./configs/zero1.yaml \
  --main_process_port=20104 \
  train_verifier.py \
  --model_name_or_path ${checkpoint_dir} \
  --data_dir data/gsm8k/model_generation \
  --target_set train \
  --save_dir ${save_dir} \
  --generator_id ${generator_id} \
  --dedup True \
  --per_problem_sampling_solution ${n_solution} \
  --loss_level token \
  --loss_on_llm True \
  --num_train_epoches 50 \
  --per_device_train_batch_size 64 \
  --per_device_eval_batch_size 64 \
  --gradient_accumulation_steps 1 \
  --gradient_checkpointing True \
  --learning_rate 1e-5 \
  --weight_decay 0 \
  --lr_scheduler_type "linear" \
  --warmup_steps 0 \
  --save_best False \
  --logging_steps 1 \
  --seed 42
```

Then the verifier model seems like working well on the training set:
![image](https://github.com/FreedomIntelligence/OVM/assets/32697489/d34103e5-50ba-4647-9b66-8a28fdd59277)

However, when I evaluate the Mistral-A (generator) and Mistral-B (verifier) using the script "eval_step_beam.sh", the accuracy score reported is only around 52.39%. I'm not sure which step is incorrect here that leads to this unexpected performance, so I did the following to check each step:

1. I also downloaded the provided verifier checkpoint (call it Mistral-C) and evaluated the beam search with Mistral-A and Mistral-C, and the accuracy score is around 81.43%, which is very close to the score reported in the paper.  I think that indicates at least the trained generator (Mistral-A) is fine and the script "eval_step_beam.sh" that I'm using to evaluate the beam search is also ok.
2.  Then I finetuned the Mistral-C using the same script (train_verifier.sh), at the beginning of the training stage, the loss of Mistral-C is close to the loss of Mistral-B in the end. Let's say this finetuned from the checkpoint verifier is Mistral-D. I then evaluated the beam search with Mistral-A and Mistral-D, the accuracy score dropped to 59.97%. I think this may mean that i mistakenly used the verifier training script or the training data, or some hyperparameters I set are wrong.  So i included the dataset link and the script i was using above, in case if it's useful.
![image](https://github.com/FreedomIntelligence/OVM/assets/32697489/5b91f39a-266c-4e96-bfe4-8ee147d497bb)

Many thanks in advance!

**Comment by OakYU** [comment]
From the script you provided, I think it might be overfitting: `--num_train_epoches 50` is a very large number, while the lr scheduler is kept unchanged (this caused the model to be trained with a large lr in many epochs). You can first try a small number of epochs, e.g. 1-3. 

Hope it can help

**Comment by chenhan97** [comment]
> From the script you provided, I think it might be overfitting: `--num_train_epoches 50` is a very large number, while the lr scheduler is kept unchanged (this caused the model to be trained with a large lr in many epochs). You can first try a small number of epochs, e.g. 1-3.
> 
> Hope it can help

Hi, thank you for your reply! I reran the whole process and it turns out i used the wrong dataset (i was using the sample dataset instead of the one from hugginface dataset https://huggingface.co/datasets/FreedomIntelligence/OVM-dataset/tree/main/gsm8k/model_generation/train ) to train the verifier. Now I trained the verifier model with just one epoch and it yields 71.9% accuracy with the generator, which is a significantly improved result. But it still has a big gap compared with the verifier downloaded from the checkpoint (81.05% with the same generator). I pasted the hyperparameter settings for generator training and verifier training here. Is it possible that some parameters i set are not efficient? 

train_generator.sh
```
accelerate launch \
  --config_file ./configs/zero1.yaml \
  --main_process_port=20650 \
  train_generator.py \
  --model_name_or_path ${model_name_or_path} \
  --dataset gsm8k \
  --data_dir data/gsm8k \
  --target_set train \
  --save_dir ${save_dir} \
  --num_train_epoches 2 \
  --eval_steps 200 \
  --per_device_train_batch_size 16 \
  --per_device_eval_batch_size 4 \
  --gradient_accumulation_steps 1 \
  --gradient_checkpointing True \
  --learning_rate 1e-5 \
  --weight_decay 0 \
  --lr_scheduler_type "linear" \
  --warmup_steps 0 \
  --save_steps 200 \
  --save_best False \
  --save_total_limit 0 \
  --logging_dir ./wandb \
  --logging_steps 8 \
  --seed 42
```

train_verifier.sh
```
n_solution=100
accelerate launch \
  --config_file ./configs/zero1.yaml \
  --main_process_port=20104 \
  train_verifier.py \
  --model_name_or_path ${checkpoint_dir} \
  --data_dir data/gsm8k/model_generation \
  --target_set train \
  --save_dir ${save_dir} \
  --generator_id ${generator_id} \
  --dedup True \
  --per_problem_sampling_solution ${n_solution} \
  --loss_level token \
  --loss_on_llm True \
  --num_train_epoches 1 \
  --eval_steps 1000 \
  --per_device_train_batch_size 64 \
  --per_device_eval_batch_size 64 \
  --gradient_accumulation_steps 1 \
  --gradient_checkpointing True \
  --learning_rate 1e-5 \
  --weight_decay 0 \
  --lr_scheduler_type "linear" \
  --warmup_steps 0 \
  --save_epoches 1 \
  --save_best False \
  --save_total_limit 0 \
  --logging_dir ./wandb \
  --logging_steps 20 \
  --seed 42
```

and configs/zero1.yaml:
```
command_file: null
commands: null
compute_environment: LOCAL_MACHINE
deepspeed_config:
  deepspeed_multinode_launcher: standard
  gradient_clipping: 1.0
  zero_stage: 1
distributed_type: DEEPSPEED
downcast_bf16: true
dynamo_backend: 'NO'
gpu_ids: null
machine_rank: 0
main_process_ip: null
main_process_port: null
main_training_function: main
fsdp_config: {}
megatron_lm_config: {}
mixed_precision: bf16
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
tpu_name: null
tpu_zone: null
use_cpu: false
```

**Comment by OakYU** [comment]
our OVM was trained with 8GPUs (64batch size * 8cards = 512 batch size in total). now you are using 4GPUs, so I recommend to set `--gradient_accumulation_steps 2` to keep the total batch size as the same. Hope it can help

**Comment by chenhan97** [comment]
> our OVM was trained with 8GPUs (64batch size * 8cards = 512 batch size in total). now you are using 4GPUs, so I recommend to set `--gradient_accumulation_steps 2` to keep the total batch size as the same. Hope it can help

Thank you for your help! I‘ve conducted some small experiments following your suggestions. I trained a verifier V1 with two epochs and set the gradient accumulation steps as 2. The trained verifier V1 with the generator achieves 71.6% accuracy (basically the same as the previous one). Then I checked the test loss of this V1 on this dataset https://huggingface.co/datasets/FreedomIntelligence/OVM-dataset/blob/main/gsm8k/model_generation/test/responses_n300_mistral7b-ep2.jsonl , which yields a 23.39 loss value on average. I also ran the loaded checkpoint verifier V2 on the same test set and obtained a 28.49 loss value (even higher than the V1 verifier).  

I'm wondering why V1 still cannot achieve 80+% accuracy while it even has lower test loss on the test set.

**Comment by OakYU** [comment]
will try to capture your problem and give you a response before next week

**Comment by OakYU** [comment]
found the problem. learning rate is 2e-6 for mistral verifier in our experiments (1e-5 is for llama), which is also mentioned in the paper. just set `--learning_rate 2e-6`, `--num_train_epoches 1`, and `--gradient_accumulation_steps 2`

**Comment by OakYU** [comment]
any progress here? I conducted experiments with our model and reproduced the results in the setting of 4gpus: lr1e-5 led to 70%+ accuracy, while lr2e-6 resulted in 80%+. if there is still a problem, feel free to let me know

**Comment by OakYU** [comment]
> > our OVM was trained with 8GPUs (64batch size * 8cards = 512 batch size in total). now you are using 4GPUs, so I recommend to set `--gradient_accumulation_steps 2` to keep the total batch size as the same. Hope it can help
> 
> Thank you for your help! I‘ve conducted some small experiments following your suggestions. I trained a verifier V1 with two epochs and set the gradient accumulation steps as 2. The trained verifier V1 with the generator achieves 71.6% accuracy (basically the same as the previous one). Then I checked the test loss of this V1 on this dataset https://huggingface.co/datasets/FreedomIntelligence/OVM-dataset/blob/main/gsm8k/model_generation/test/responses_n300_mistral7b-ep2.jsonl , which yields a 23.39 loss value on average. I also ran the loaded checkpoint verifier V2 on the same test set and obtained a 28.49 loss value (even higher than the V1 verifier).
> 
> I'm wondering why V1 still cannot achieve 80+% accuracy while it even has lower test loss on the test set.

about the mismatching between loss and accuracy: loss only reflects the prediction at the token level. but when we conduct guided decoding, we need to compare scores among multiple samples, which concerns the relative comparison (i.e. higher/lower). so the loss is less correlated to accuracy here. let's image that, if each step appears only once in the whole dataset, the lowest loss is achieved by predicting all the steps leading to the correct answers as 1, otherwise 0. but when it comes to guided decoding, we can't determine the most promising steps if all the potential steps' values are predicted as 1. the algorithm will degrade to randomly sampling among potential steps (filter these predicted as 0). so in this case, optimal loss, but suboptimal accuracy

**Comment by JianqiaoLu** [comment]
Thank you for sharing, can you share what is the final convergence loss value v_loss of the Mistral verifier after training for 2 epochs？ an approximate range is appreciated since I know many factors contribute to the final v_loss.

**Comment by OakYU** [comment]
> Thank you for sharing, can you share what is the final convergence loss value v_loss of the Mistral verifier after training for 2 epochs？ an approximate range is appreciated since I know many factors contribute to the final v_loss.

In my experiments, the v_loss is 5-20 when training mistral with lr1e-5, 2 epochs, 4 A100 cards, batch size 64, and gradient accumulation steps 2. However, the learning rate of this setting is inappropriate, leading to a suboptimal final performance.

In our paper, we train mistral with lr2e-6 in 1 epoch, the final v_loss is also 5-20
